_posts/2015-09-12-developing-gamepad-virtual-keyboard.markdown

```diff
@@ -0,0 +1,150 @@
+---
+layout: post
+title: "Considerations developing a gamepad virtual keyboard, Part 1: The interface"
+tags: flux daisywheeljs gamepad html5
+username: likethemammal
+---
+
+When working with gamepad controllers, it's general practice to present users with a virtual keyboard to enter text information. Keyboard layouts like [DaisywheelJS’s](http://daisywheeljs.org) and [others](http://www.inference.phy.cam.ac.uk/dasher/images/newdasher.gif) are great alternatives to the traditional QWERTY keyboard presented by console developers. But whether you’re working with an unique layout or not, there are several considerations that go into creating a virtual keyboard. This article discusses those topics. This first part will cover the visual aspects (the input element, keyboard, controls, etc.), and the next part will cover connecting everything to a gamepad (in this case, the HTML5 Gamepad API).
+
+*Note: When I wrote DaisywheelJS I used React and Flux for the keyboard’s logic and view layer. Because of this, I’ll be loosely sprinkling bits of Flux-specific implementation and some ignorable code examples. Most of the article will be a discussion of technology-agnostic considerations, so don’t worry if you don’t use Flux.*
+
+![Zelda's virtual keyboard](http://i27.fastpic.ru/big/2011/1003/a2/1e6d3aa24b1f36fc471a764dae939da2.jpg)
+
+To start off, here are the common pieces of a virtual keyboard interface:
+
+ + **The keyboard layout**: Shows available characters and the currently selected character.
+ + **The input fields**: Used to show what the user is typing and retain the text when they're done typing.
+ + **The Viewport**: The container around everything.
+ + **The controls**: A list of the gamepad controls available to the user.
+
+Each of these elements will be covered individually.
+
+### The keyboard layout
+
+This is the part I’ll talk about the least, because implementation varies greatly with the type of layout. The layout could be a radial keyboard, like the Daisywheel, a traditional keyboard, like QWERTY, or something completely different like [Beyond Good and Evil’s keyboard](https://youtu.be/euquOpUmUyk)). No matter what layout you have, let’s assume the user can only type one character at a time. This could be accessed by some sort of event or callback that sends the symbol to any part of the app that’s interested. One of those parts is the Input field.
```

Comment:
The layout could be a radial keyboard (like the Daisywheel), a traditional keyboard (like QWERTY), or something completely different (like [Beyond Good and Evil’s keyboard](https://youtu.be/euquOpUmUyk))